File: app/src/main/java/com/icerockdev/babenko/ui/data_fields/DataFieldsInteractorImpl.java
package com.icerockdev.babenko.ui.data_fields;

import android.support.annotation.NonNull;
import android.support.v4.util.SparseArrayCompat;
import android.util.Patterns;
import android.widget.EditText;

import com.icerockdev.babenko.model.entities.DataField;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

import io.reactivex.Single;

import static com.icerockdev.babenko.core.ApplicationConstants.EMAIL;
import static com.icerockdev.babenko.core.ApplicationConstants.NUMBER;
import static com.icerockdev.babenko.core.ApplicationConstants.PHONE;
import static com.icerockdev.babenko.core.ApplicationConstants.TEXT;
import static com.icerockdev.babenko.core.ApplicationConstants.URL;

/**
 * Created by Roman Babenko on 30/04/17.
 */

public class DataFieldsInteractorImpl implements DataFieldsInteractor {

    private DataField[] mDataFields;

    public DataFieldsInteractorImpl(@NonNull DataField[] dataFields) {
        mDataFields = dataFields;
    }

    public Single<List<Integer>> checkFields(SparseArrayCompat<EditText> fieldValues) {
        return Single.fromCallable(() -> {
            List<Integer> errorList = new ArrayList<Integer>();
            for (int i = 0; i < fieldValues.size(); i++) {
                int key = fieldValues.keyAt(i);
                EditText fieldValue = fieldValues.get(key);
                for (DataField dataField : mDataFields)
                    if (dataField.getId() == key) {
                        if (!isDataFieldCorrect(fieldValue.getText().toString(), dataField.getType()))
                            errorList.add(key);
                    }
            }
            return errorList;
        });
    }

    private boolean isDataFieldCorrect(String data, String type) {
        if (data.isEmpty())
            return false;
        switch (type) {
            case TEXT:
                int length = data.length();
                return (length > 10 && length < 30);
            case EMAIL:
                return android.util.Patterns.EMAIL_ADDRESS.matcher(data).matches();
            case PHONE:
                Pattern phonePattern = Pattern.compile("^[+7]{2}\\d{10}$");
                return phonePattern.matcher(data).matches();
            case NUMBER:
                Pattern numberPattern = Pattern.compile("^\\d{1,5}$");
                return numberPattern.matcher(data).matches();
            case URL:
                return Patterns.WEB_URL.matcher(data).matches();
            default:
                throw new IllegalArgumentException("Unknown type");
        }
    }

    public Single<List<DataField>> requestDataFields() {
        return Single.fromCallable(() -> Arrays.asList(mDataFields));
    }
}

File: app/src/main/java/com/icerockdev/babenko/ui/data_fields/DataFieldsInteractorImpl.kt
package com.icerockdev.babenko.ui.data_fields

import android.support.v4.util.SparseArrayCompat
import android.util.Patterns
import android.widget.EditText
import com.icerockdev.babenko.core.ApplicationConstants.EMAIL
import com.icerockdev.babenko.core.ApplicationConstants.NUMBER
import com.icerockdev.babenko.core.ApplicationConstants.PHONE
import com.icerockdev.babenko.core.ApplicationConstants.TEXT
import com.icerockdev.babenko.core.ApplicationConstants.URL
import com.icerockdev.babenko.model.entities.DataField
import io.reactivex.Single
import java.util.*
import java.util.regex.Pattern

/**
 * Created by Roman Babenko on 30/04/17.
 */

class DataFieldsInteractorImpl(private val mDataFields: Array<DataField>) : DataFieldsInteractor {
    override fun checkFields(fieldValues: SparseArrayCompat<EditText>): Single<List<Int>> {
        return Single.fromCallable {
            val errorList = ArrayList<Int>()
            for (i in 0 until fieldValues.size()) {
                val key = fieldValues.keyAt(i)
                val fieldValue = fieldValues.get(key)
                for (dataField in mDataFields)
                    if (dataField.id == key) {
                        if (!isDataFieldCorrect(fieldValue.text.toString(), dataField.type))
                            errorList.add(key)
                    }
            }
            errorList
        }
    }

    private fun isDataFieldCorrect(data: String, type: String?): Boolean {
        if (data.isEmpty())
            return false
        when (type) {
            TEXT -> {
                val length = data.length
                return length in 11..29
            }
            EMAIL -> return android.util.Patterns.EMAIL_ADDRESS.matcher(data).matches()
            PHONE -> {
                val phonePattern = Pattern.compile(phoneRegex)
                return phonePattern.matcher(data).matches()
            }
            NUMBER -> {
                val numberPattern = Pattern.compile(numberRegex)
                return numberPattern.matcher(data).matches()
            }
            URL -> {
                return Patterns.WEB_URL.matcher(data).matches()
            }
            else -> {
                throw IllegalArgumentException("Unknown type")
            }
        }
    }

    override fun requestDataFields(): Single<List<DataField>> {
        return Single.fromCallable { Arrays.asList(*mDataFields) }
    }

    companion object {
        const val phoneRegex = "^[+7]{2}\\d{10}$"
        const val numberRegex = "^\\d{1,5}$"
    }
}
